### PR TITLE
unittests for python API

### PR DESCRIPTION
**Pull request by CFGrote**
Related to #32 

This PR introduces unittesting of the python wrappers. Only new dependency is the [unittest library](https://docs.python.org/3/library/unittest.html) which comes with the standard python libraries. Some tests use the sample files that have to be pulled with the .travis/download_samples.sh script. 

To run the tests, do `python test/python/unittest/Test.py -v`. To add more tests, use `tests/python/unittest/API/APITest.py` as a template and don't forget to import new test modules `Test.py` and to add the corresponding test suite. I'll be happy to assist.

### CODE REVIEW

**Comment by ax3l** [comment]
@CFGrote hurray, thank you so much for the PR, Carsten!

Can you please add a PR description? Especially, can you highlight what dependencies are needed for the tests, e.g. are they part of the python standard libraries or external? How to call the tests manually, how to add more tests (e.g. in separate files?), etc.?

File: include/openPMD/Mesh.hpp
Comment (ax3l) [inline include/openPMD/Mesh.hpp]: Please revert the addition in this file since it's a bit outdated.
Records and record components are accessed like this instead:
https://github.com/openPMD/openPMD-api/blob/0.2.0-alpha/examples/2_read_serial.py#L34-L44

File: src/binding/python/Mesh.cpp
Comment (ax3l) [inline src/binding/python/Mesh.cpp]: same here, this should not be needed to be exposed

File: src/binding/python/Mesh.cpp
Comment (ax3l) [inline src/binding/python/Mesh.cpp]: same here, this should not be needed

File: test/python/unittest/API/APITest.py
Comment (ax3l) [inline test/python/unittest/API/APITest.py]: Should be accessible via:
```python
record = self.__series.iterations[100].particles['electrons']['positions']['x']
```

File: test/python/unittest/API/APITest.py
Comment (ax3l) [inline test/python/unittest/API/APITest.py]: you can do that now, see the example here :-)
https://github.com/openPMD/openPMD-api/blob/0.2.0-alpha/examples/2_read_serial.py#L34-L44

**Review by ax3l — DISMISSED** [review]
Very nice draft!
I added some inline comments on already implemented parts that can be used.

**Comment by CFGrote** [comment]
@ax3l can you add to .travis.yml that these test actrually run ? sth. like python test/unittest/Test.py -v should do

**Comment by ax3l** [comment]
Just as a note, I am right now fixing a CI issue with HDF5. Please rebase your branch against `dev` as soon as we merged #250

Yes, I can push on your branch to address CMake and CI related features in the end :-)

**Comment by ax3l** [comment]
@CFGrote CI fixed, please fetch and rebase :-)
edit: never mind, my manual restart fixes this as well and merged onto the new dev.

File: test/python/unittest/API/APITest.py
Comment (ax3l) [inline test/python/unittest/API/APITest.py]: can you try if

```python
record = self.__series.iterations[100].particles['electrons']['position']['x']
```
works?

File: test/python/unittest/API/APITest.py
Comment (ax3l) [inline test/python/unittest/API/APITest.py]: works now, please see example in https://github.com/openPMD/openPMD-api/blob/0.2.0-alpha/examples/2_read_serial.py#L34-L47

**Review by ax3l — CHANGES_REQUESTED** [review]
Forgotten inline comments re-added :-)

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py]: This can be removed together with the C++ [code](https://github.com/openPMD/openPMD-api/blob/06ae64508f726d18eaedc60d7e73580e7f24a054/include/openPMD/RecordComponent.hpp#L52-L57) and [bindings](https://github.com/openPMD/openPMD-api/blob/b95e33c6627aa41afca80bad0d2265846451f5e4/src/binding/python/RecordComponent.cpp#L178-L182) behind it. We do not allow the user to specify allocation strategy any more.
If you include this, I'll remove it when we get rid of the ```enum``` with #253.

File: test/python/unittest/Test.py
Comment (ax3l) [inline test/python/unittest/Test.py]: please add the same copyright line as in the other file :)

File: test/python/unittest/TestUtilities/TestUtilities.py
Comment (ax3l) [inline test/python/unittest/TestUtilities/TestUtilities.py]: please here as well

File: test/python/unittest/API/APITest.py
Comment (CFGrote) [reply test/python/unittest/API/APITest.py]:
```
ERROR: testRecord (__main__.APITest)
Test openPMD.Record.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "APITest.py", line 293, in testRecord
    record = self.__series.iterations[100].particles['electrons']['positions']['x']
TypeError: 'openPMD.ParticleSpecies' object is not subscriptable

```

File: test/python/unittest/API/APITest.py
Comment (CFGrote) [reply test/python/unittest/API/APITest.py]: I'll leave it in but modify: now asserts AttributeError on openPMD.Allocation .

File: test/python/unittest/Test.py
Comment (CFGrote) [reply test/python/unittest/Test.py]: done

File: test/python/unittest/TestUtilities/TestUtilities.py
Comment (CFGrote) [reply test/python/unittest/TestUtilities/TestUtilities.py]: done

File: test/python/unittest/API/APITest.py
Comment (CFGrote) [reply test/python/unittest/API/APITest.py]: ok, works. test added

**Comment by ax3l** [comment]
@CFGrote I think the rebase includes a little too many commits now. Just rebase on mainline's `dev` (fetch first) and remove the ones not coming from you on top :-)

File: test/python/unittest/API/APITest.py
Comment (ax3l) [reply test/python/unittest/API/APITest.py]: Ah, alright. Not implemented yet, a `FIXME` note is great.

If you rebase against the latest `dev`, the particle access should work up to:
```python
self.__series.iterations[100].particles['electrons']['position']
```

and still lacks access to the *record component* `['x']`

**Comment by ax3l** [comment]
@CFGrote can I help somehow to finish this PR? :-)

**Comment by CFGrote** [comment]
i fear i f***ed up my repo. @ax3l, can you solve the conflict from your end?

**Review by ax3l — APPROVED** [review]

**Comment by ax3l** [comment]
@CFGrote shall I fix the PEP issues and add CI?
(that means I will directly push on your PR branch, so don't forget to fetch & pull later on. Also, I think you forgot to branch before the PR ;-) )

**Review by ax3l — APPROVED** [review]

**Comment by ax3l** [comment]
@C0nsultant feel free to review & merge this if all passes :)

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py]: To suppress all warnings
> Series constructor called with explicit iteration suggests loading a single file with groupBased iteration encoding. Loaded file is fileBased.

plase replace the explicit iteration numbers in these filenames (i.e. ```00000400``` and ```00000100```) with the [iterationFormat](https://github.com/openPMD/openPMD-standard/blob/1.1.0/STANDARD.md#iterations-and-time-series) regex ```%T```.
This will still load your desired iterations (additionally to all other iterations in that Series).

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py]: If the ```print()``` statements are used for file inspection, replace them with assertions where possible.
Reflecting the logic from the Python examples may be helpful during development of test cases, but only adds noise when running the tests.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py]: This is a good candidate for a ```testContainer``` case where you check if 
```python
iter(series.iterations)
```
returns a sane result.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py]: Before accessing with ```[]```, you should assert that Iteration 400 is actually contained. This API creates objects (with sane openPMD defaults) for missing keys instead of raising ```KeyError```.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py]: Iteration number is hardcoded.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py]:
```python
# Get a particle species.
```

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py:168]: This is reduant in the current form. The one contained assertion is included in the other three Series test cases.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py:161]: This should be moved into ```testRecord_Component```.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py:166]: The shape, datatype and content of the read chunk should be verified.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py]: This is now done in the loop above. Maybe we should check if this succeeds for any newly created Iteration
```python
i not in series.iterations
```

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [reply test/python/unittest/API/APITest.py]: In that testcase, we should also check what happens when you access invalid keys.
```python
series.iterations['definitely not an int']
```
In C++, this is caught by the compiler. I'm not sure about the result in Python.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py]: The datatype and content of contained attributes should be verified in all testcases that test concrete openPMD objects.

Comment (anokfireball) [reply test/python/unittest/API/APITest.py]: Getters and setters for attributes required by the standard should also be verified.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py]: @ax3l does this return deep copies or are these shallow copies as intended with #144?
Assuming it is a shallow copy, we should check if
 - the contained elements and attributes are the same and 
 - reflect modification in the other handle 

(as well as in all other objects with this behaviour).

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py]: This is a mis-design in the C++ API. ```Container``` should not have a publicly available constructor. #280

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py:312]: A ```Mesh``` is a ```Record``` by [design](https://github.com/openPMD/openPMD-standard/blob/1.1.0/STANDARD.md#mesh-based-records). What this actually tests are ```Mesh_Record_Component```s.

File: test/python/unittest/TestUtilities/TestUtilities.py
Comment (anokfireball) [inline test/python/unittest/TestUtilities/TestUtilities.py]: Hard-coded directory separators defeat the purpose of using ```os.path.join```.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py:9]: This should be moved to the third-party imports.

**Review by anokfireball — CHANGES_REQUESTED** [review]
I noted a lot of stuff. Do not despair, you don't need to do all of it.
The annotations are there so any leftovers can be done in a follow-up.

The iterationFormat regex and the ```print()``` noise are a blocker, though.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [reply test/python/unittest/API/APITest.py]: This should be fine as the Series is opened with ```openPMD.Access_Type.read_only```.

File: test/python/unittest/API/APITest.py
Comment (ax3l) [reply test/python/unittest/API/APITest.py]: In python I throw on that as soon as the series is read_only. That's also the place where it originally triggered me that we need an asset in C++ as well :)

**Comment by ax3l** [comment]
@C0nsultant I tried to address a whole bunch of review comments, but could not yet fix all.
`print()`s are removed, paths fixed and more verifies added. If you like, we could merge the current state and improve it ourselves further?

File: test/python/unittest/API/APITest.py
Comment (ax3l) [reply test/python/unittest/API/APITest.py]: I have to check, I hope it's a reference in the first access and a shallow-copy / alias in the second case.

Comment (ax3l) [reply test/python/unittest/API/APITest.py]: That's not an ideal place to check, here it uses a readonly series which means I can't modify the iteration for proper inspection :S

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py:18]: Ah yes, a simple runtime ```try```-```except``` to turn off features instead of CMake magic with ```#ifdef``` conditional compilation. :slightly_smiling_face:

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py:45]: Ideally, these paths would be created with ```os.path.join``` as well.

File: test/python/unittest/API/APITest.py
Comment (anokfireball) [inline test/python/unittest/API/APITest.py:91]: This does not use the pure particle dataset referred to in the docstring.

**Review by anokfireball — APPROVED** [review]

**Comment by anokfireball** [comment]
Thank you very much, @CFGrote!

**Comment by ax3l** [comment]
What could possibly go wrong... ;)

File: test/python/unittest/API/APITest.py
Comment (ax3l) [reply test/python/unittest/API/APITest.py:18]: What could possibly go wrong... ;)

**Pull request merged**
